// @mx-space/api-client@0.10.6/lib/controllers/aggregate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateController = void 0;
const utils_1 = require("../utils");
const auto_bind_1 = require("../utils/auto-bind");
class AggregateController {
    constructor(client) {
        this.client = client;
        this.base = 'aggregate';
        this.name = 'aggregate';
        (0, auto_bind_1.autoBind)(this);
    }
    get proxy() {
        return this.client.proxy(this.base);
    }
    /**
     * 获取聚合数据
     */
    getAggregateData() {
        return this.proxy.get();
    }
    /**
     * 获取最新发布的内容
     */
    getTop(size = 5) {
        return this.proxy.top.get({ params: { size } });
    }
    getTimeline(options) {
        const { sort, type, year } = options || {};
        return this.proxy.timeline.get({
            params: {
                sort: sort && (0, utils_1.sortOrderToNumber)(sort),
                type,
                year,
            },
        });
    }
    /**
     * 获取聚合数据统计
     */
    getStat() {
        return this.proxy.stat.get();
    }
}
exports.AggregateController = AggregateController;
