// @mx-space/api-client@0.10.6/lib/controllers/category.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CategoryController = void 0;
const utils_1 = require("../utils");
const auto_bind_1 = require("../utils/auto-bind");
const error_1 = require("../core/error");
const category_1 = require("../models/category");
class CategoryController {
    constructor(client) {
        this.client = client;
        this.name = 'category';
        this.base = 'categories';
        (0, auto_bind_1.autoBind)(this);
    }
    get proxy() {
        return this.client.proxy(this.base);
    }
    getAllCategories() {
        return this.proxy.get({
            params: {
                type: category_1.CategoryType.Category,
            },
        });
    }
    getAllTags() {
        return this.proxy.get({
            params: {
                type: category_1.CategoryType.Tag,
            },
        });
    }
    async getCategoryDetail(ids) {
        if (typeof ids === 'string') {
            const data = await this.proxy.get({
                params: {
                    ids,
                },
            });
            const result = Object.values(data.entries)[0];
            (0, utils_1.attachRawFromOneToAnthor)(data, result);
            return result;
        }
        else if (Array.isArray(ids)) {
            const data = await this.proxy.get({
                params: {
                    ids: ids.join(','),
                },
            });
            const entries = data?.entries;
            if (!entries) {
                throw new error_1.RequestError('data structure error', 500, data.$request.path, data);
            }
            const map = new Map(Object.entries(entries).map(([id, value]) => [id.toLowerCase(), value]));
            (0, utils_1.attachRawFromOneToAnthor)(data, map);
            return map;
        }
    }
    async getCategoryByIdOrSlug(idOrSlug) {
        const res = await this.proxy(idOrSlug).get();
        return (0, utils_1.destructureData)(res);
    }
    async getTagByName(name) {
        const res = await this.proxy(name).get({
            params: {
                tag: 1,
            },
        });
        return res;
    }
}
exports.CategoryController = CategoryController;
