// @mx-space/api-client@0.10.6/lib/controllers/comment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommentController = void 0;
const auto_bind_1 = require("../utils/auto-bind");
class CommentController {
    constructor(client) {
        this.client = client;
        this.base = 'comments';
        this.name = 'comment';
        (0, auto_bind_1.autoBind)(this);
    }
    get proxy() {
        return this.client.proxy(this.base);
    }
    /**
     * 根据 comment id 获取评论, 包括子评论
     */
    getById(id) {
        return this.proxy(id).get();
    }
    /**
     * 获取文章的评论列表
     * @param refId 文章 Id
     */
    getByRefId(refId, pagination = {}) {
        const { page, size } = pagination;
        return this.proxy
            .ref(refId)
            .get({
            params: { page: page || 1, size: size || 10 },
        });
    }
    /**
     * 评论
     */
    comment(refId, data) {
        return this.proxy(refId).post({
            data,
        });
    }
    /**
     * 回复评论
     */
    reply(commentId, data) {
        return this.proxy.reply(commentId).post({
            data,
        });
    }
}
exports.CommentController = CommentController;
