// @mx-space/api-client@0.10.6/lib/controllers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicController = exports.UserController = exports.ServerlessController = exports.SnippetController = exports.SearchController = exports.SayController = exports.RecentlyController = exports.ProjectController = exports.PostController = exports.PageController = exports.NoteController = exports.LinkController = exports.CommentController = exports.CategoryController = exports.AggregateController = exports.allContollerNames = exports.allControllers = void 0;
const aggregate_1 = require("./aggregate");
Object.defineProperty(exports, "AggregateController", { enumerable: true, get: function () { return aggregate_1.AggregateController; } });
const category_1 = require("./category");
Object.defineProperty(exports, "CategoryController", { enumerable: true, get: function () { return category_1.CategoryController; } });
const comment_1 = require("./comment");
Object.defineProperty(exports, "CommentController", { enumerable: true, get: function () { return comment_1.CommentController; } });
const link_1 = require("./link");
Object.defineProperty(exports, "LinkController", { enumerable: true, get: function () { return link_1.LinkController; } });
const note_1 = require("./note");
Object.defineProperty(exports, "NoteController", { enumerable: true, get: function () { return note_1.NoteController; } });
const page_1 = require("./page");
Object.defineProperty(exports, "PageController", { enumerable: true, get: function () { return page_1.PageController; } });
const post_1 = require("./post");
Object.defineProperty(exports, "PostController", { enumerable: true, get: function () { return post_1.PostController; } });
const project_1 = require("./project");
Object.defineProperty(exports, "ProjectController", { enumerable: true, get: function () { return project_1.ProjectController; } });
const recently_1 = require("./recently");
Object.defineProperty(exports, "RecentlyController", { enumerable: true, get: function () { return recently_1.RecentlyController; } });
const say_1 = require("./say");
Object.defineProperty(exports, "SayController", { enumerable: true, get: function () { return say_1.SayController; } });
const search_1 = require("./search");
Object.defineProperty(exports, "SearchController", { enumerable: true, get: function () { return search_1.SearchController; } });
const severless_1 = require("./severless");
Object.defineProperty(exports, "ServerlessController", { enumerable: true, get: function () { return severless_1.ServerlessController; } });
const snippet_1 = require("./snippet");
Object.defineProperty(exports, "SnippetController", { enumerable: true, get: function () { return snippet_1.SnippetController; } });
const topic_1 = require("./topic");
Object.defineProperty(exports, "TopicController", { enumerable: true, get: function () { return topic_1.TopicController; } });
const user_1 = require("./user");
Object.defineProperty(exports, "UserController", { enumerable: true, get: function () { return user_1.UserController; } });
exports.allControllers = [
    aggregate_1.AggregateController,
    category_1.CategoryController,
    comment_1.CommentController,
    link_1.LinkController,
    note_1.NoteController,
    page_1.PageController,
    post_1.PostController,
    project_1.ProjectController,
    recently_1.RecentlyController,
    topic_1.TopicController,
    say_1.SayController,
    search_1.SearchController,
    snippet_1.SnippetController,
    severless_1.ServerlessController,
    user_1.UserController,
];
exports.allContollerNames = [
    'aggregate',
    'category',
    'comment',
    'link',
    'note',
    'page',
    'post',
    'project',
    'topic',
    'recently',
    'say',
    'search',
    'snippet',
    'serverless',
    'user',
    // alias,
    'friend',
    'master',
    'shorthand',
];
