// @mx-space/api-client@0.10.6/lib/controllers/page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageController = void 0;
const auto_bind_1 = require("../utils/auto-bind");
class PageController {
    constructor(client) {
        this.client = client;
        this.base = 'pages';
        this.name = 'page';
        (0, auto_bind_1.autoBind)(this);
    }
    get proxy() {
        return this.client.proxy(this.base);
    }
    /**
     * 页面列表
     */
    getList(page = 1, perPage = 10, options = {}) {
        const { select, sortBy, sortOrder } = options;
        return this.proxy.get({
            params: {
                page,
                size: perPage,
                select: select?.join(' '),
                sortBy,
                sortOrder,
            },
        });
    }
    /**
     * 页面详情
     */
    getById(id) {
        return this.proxy(id).get();
    }
    /**
     * 根据路径获取页面
     * @param slug 路径
     * @returns
     */
    getBySlug(slug) {
        return this.proxy.slug(slug).get({});
    }
}
exports.PageController = PageController;
