// @mx-space/api-client@0.10.6/lib/controllers/post.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostController = void 0;
const auto_bind_1 = require("../utils/auto-bind");
class PostController {
    constructor(client) {
        this.client = client;
        this.base = 'posts';
        this.name = 'post';
        (0, auto_bind_1.autoBind)(this);
    }
    get proxy() {
        return this.client.proxy(this.base);
    }
    /**
     * 获取文章列表分页
     * @param page
     * @param perPage
     * @returns
     */
    getList(page = 1, perPage = 10, options = {}) {
        const { select, sortBy, sortOrder, year } = options;
        return this.proxy.get({
            params: {
                page,
                size: perPage,
                select: select?.join(' '),
                sortBy,
                sortOrder,
                year,
            },
        });
    }
    getPost(idOrCategoryName, slug) {
        if (arguments.length == 1) {
            return this.proxy(idOrCategoryName).get();
        }
        else {
            return this.proxy(idOrCategoryName)(slug).get();
        }
    }
    /**
     * 获取最新的文章
     */
    getLatest() {
        return this.proxy.latest.get();
    }
    /**
     * 点赞
     */
    thumbsUp(id) {
        return this.proxy('_thumbs-up').get({ params: { id } });
    }
}
exports.PostController = PostController;
