// @mx-space/api-client@0.10.6/lib/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attachRawFromOneToAnthor = exports.destructureData = exports.sortOrderToNumber = exports.isPlainObject = void 0;
const isPlainObject = (obj) => isObject(obj) &&
    Object.prototype.toString.call(obj) === '[object Object]' &&
    Object.getPrototypeOf(obj) === Object.prototype;
exports.isPlainObject = isPlainObject;
const sortOrderToNumber = (order) => {
    return ({
        asc: 1,
        desc: -1,
    }[order] || 1);
};
exports.sortOrderToNumber = sortOrderToNumber;
const isObject = (obj) => obj && typeof obj === 'object';
const destructureData = (payload) => {
    if (typeof payload !== 'object') {
        return payload;
    }
    if (payload === null) {
        return payload;
    }
    const data = payload.data;
    const dataIsPlainObject = (0, exports.isPlainObject)(data);
    if (dataIsPlainObject && Object.keys(payload).length === 1) {
        const d = Object.assign({}, data);
        // attach raw onto new data
        (0, exports.attachRawFromOneToAnthor)(payload, d);
        return d;
    }
    return payload;
};
exports.destructureData = destructureData;
const attachRawFromOneToAnthor = (from, to) => {
    if (!from || !isObject(to)) {
        return;
    }
    from.$raw &&
        Object.defineProperty(to, '$raw', {
            value: { ...from.$raw },
            enumerable: false,
        });
    from.$request &&
        Object.defineProperty(to, '$request', {
            value: { ...from.$request },
            enumerable: false,
        });
};
exports.attachRawFromOneToAnthor = attachRawFromOneToAnthor;
