// @mx-space/api-client@0.10.6/lib/adaptors/axios.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.axiosAdaptor = void 0;
const axios_1 = __importDefault(require("axios"));
const $http = axios_1.default.create({});
// ignore axios `method` declare not assignable to `Method`
exports.axiosAdaptor = Object.preventExtensions({
    get default() {
        return $http;
    },
    responseWrapper: {},
    get(url, options) {
        // @ts-ignore
        return $http.get(url, options);
    },
    post(url, options) {
        const { data, ...config } = options || {};
        // @ts-ignore
        return $http.post(url, data, config);
    },
    put(url, options) {
        const { data, ...config } = options || {};
        // @ts-ignore
        return $http.put(url, data, config);
    },
    delete(url, options) {
        const { ...config } = options || {};
        // @ts-ignore
        return $http.delete(url, config);
    },
    patch(url, options) {
        const { data, ...config } = options || {};
        // @ts-ignore
        return $http.patch(url, data, config);
    },
});
